client/src/main/java/io/pravega/client/netty/impl/RawClient.java

```diff
@@ -0,0 +1,141 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class RawClient implements AutoCloseable {
+
+    private final CompletableFuture<ClientConnection> connection;
+    
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private final Map<Long, CompletableFuture<Reply>> requests = new HashMap<>();
+    private final ResponseProcessor responseProcessor = new ResponseProcessor();
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+    private final class ResponseProcessor extends FailingReplyProcessor {
+        
+        @Override
+        public void process(Reply reply) {
+            if (reply instanceof Hello) {
+                Hello hello = (Hello) reply;
+                log.info("Received hello: {}", hello);
+                if (hello.getLowVersion() > WireCommands.WIRE_VERSION || hello.getHighVersion() < WireCommands.OLDEST_COMPATIBLE_VERSION) {
+                    closeConnection(new IllegalStateException("Incompatible wire protocol versions " + hello));
+                }
+            } else {
+                log.debug("Received reply {}", reply);
+                reply(reply);
+            }
+        }
+        
+        @Override
+        public void connectionDropped() {
+            closeConnection(new ConnectionFailedException());
+        }
+
+        @Override
+        public void processingFailure(Exception error) {
+            log.warn("Processing failure: ", error);
+            closeConnection(error);
+        }
+        
+        @Override
+        public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenCheckFailed) {
+            log.warn("Auth token failure: ", authTokenCheckFailed);
+            closeConnection(new AuthenticationException(authTokenCheckFailed.toString()));
+        }
+    }
+    
+    public RawClient(Controller controller, ConnectionFactory connectionFactory, Segment segmentId) {
+        connection = controller.getEndpointForSegment(segmentId.getScopedName())
+                               .thenCompose((PravegaNodeUri uri) -> connectionFactory.establishConnection(uri, responseProcessor));
+        connection.exceptionally(e -> {
+            closeConnection(e);
+            throw new CompletionException(e);
+        });
+    }
+
+    private void reply(Reply reply) {
+        CompletableFuture<Reply> future;
+        synchronized (lock) {
+            future = requests.remove(reply.getRequestId());
+        }
+        if (future != null) {
+            future.complete(reply);
+        }
+    }
+    
+    private void closeConnection(Throwable exceptionToInflightRequests) {
+        log.info("Closing connection with exception: {}", exceptionToInflightRequests.getMessage());
+        connection.thenAccept(c -> {
+            try {
+                c.close();
+            } catch (Exception e) {
+                log.warn("Exception tearing down connection: ", e);
+            }
+        });
+        closed.set(true);
```

Comment:
Do you ever query this internally? Also, should we re-attempt to close the connection if `closed` is set to `true`?

Reply:
It is not.
Done.